### ISSUE TITLE
Stream f m r as generalised vector type?

### ISSUE THREAD

**Issue by pranaysashank** [haskell-streaming/streaming#75]
Generalising the Stream type in `Data.Vector.Fusion.Stream.Monadic` to

```Haskell
data Stream f m r = forall s. Stream (s -> m (Step f s r)) s

data Step f s r = Yield !(f s)
                | Skip s
                | Return r
```

I have benchmarked a few of its operations [here](https://github.com/pranaysashank/streaming-fusion). Most of the functions are direct copies from vector, and it seems to perform as good as Vector.
Comments?

**Comment by chessai** [comment]
I'm confused. What is the motivation behind this issue? That doesn't seem very clear to me.

**Comment by chessai** [comment]
Are you asking if we'd be interested in changing the internals to match this Stream type you talk about here? If so, I've thought about it before. I believe streamly uses this and in microbenchmarks is more performant.

**Comment by pranaysashank** [comment]
Yes, this is similar to the Stream type streamly and vector uses, extended with a functor layer. I wrote it  trying to understand the internals of the streaming library. 
I was interested to know, if people have thought about it before and decided there are some operations that can't be represented using this type? As far as I can tell there doesn't seem to be such a problem, other than weaker constraints on some functions.

**Comment by pranaysashank** [comment]
This issue seems related https://github.com/haskell-streaming/streaming/issues/18
